ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs

```diff
@@ -577,44 +579,98 @@ fromList k n (anchor, signers, ebbs) =
-serializationFormatVersion0 = 0
-
-encodePBftState :: Serialise (PBftVerKeyHash c)
+toCompactFormat ::
+     forall c. PBftCrypto c
+  => PBftState c
+  -> ( Map (PBftVerKeyHash c) ([SlotNo], [SlotNo], [SlotNo], [SlotNo])
+       -- For each signer, the slots in 'preAnchor', 'postAnchor',
+       -- 'preWindow', and 'inWindow'.
+     , MaybeEbbInfo
+     )
+toCompactFormat PBftState{..} = (signerSlots, ebbs)
+  where
+    toSlots :: StrictSeq (PBftSigner c) -> Map (PBftVerKeyHash c) [SlotNo]
+    toSlots =
+        Foldable.foldl'
+          (\acc (PBftSigner slot key) -> Map.insertWith (<>) key [slot] acc)
+          Map.empty
+
+    anchorSlots :: Map (PBftVerKeyHash c) ([SlotNo], [SlotNo])
+    anchorSlots =
+        Map.mergeWithKey
+          (\_key preA postA -> Just (preA, postA))
+          (Map.map (,[]))
+          (Map.map ([],))
+          (toSlots preAnchor)
+          (toSlots postAnchor)
+
+    windowSlots :: Map (PBftVerKeyHash c) ([SlotNo], [SlotNo])
+    windowSlots =
+        Map.mergeWithKey
+          (\_key preW inW -> Just (preW, inW))
+          (Map.map (,[]))
+          (Map.map ([],))
+          (toSlots preWindow)
+          (toSlots inWindow)
+
+    signerSlots :: Map (PBftVerKeyHash c) ([SlotNo], [SlotNo], [SlotNo], [SlotNo])
+    signerSlots =
+        Map.mergeWithKey
+          (\_key (preA, postA) (preW, inW) -> Just (preA, postA, preW, inW))
+          (Map.map (\(preA, postA) -> (preA, postA, [],   [])))
+          (Map.map (\(preW, inW)   -> ([],   [],    preW, inW)))
+          anchorSlots
+          windowSlots
+
+fromCompactFormat ::
+     forall c. PBftCrypto c
+  => ( Map (PBftVerKeyHash c) ([SlotNo], [SlotNo], [SlotNo], [SlotNo])
+     , MaybeEbbInfo
+     )
+  -> PBftState c
+fromCompactFormat (signerSlots, ebbs) = PBftState {..}
+  where
+    -- We use difference list for more efficient appends.
+    dlPreA, dlPostA, dlPreW, dlInW :: [PBftSigner c] -> [PBftSigner c]
+    (dlPreA, dlPostA, dlPreW, dlInW) =
+        Map.foldlWithKey'
+          (\(dlPreA', dlPostA', dlPreW', dlInW')
+            key
+            (preA, postA, preW, inW) ->
+              ( (<> map (`PBftSigner` key) preA)  . dlPreA'
+              , (<> map (`PBftSigner` key) postA) . dlPostA'
+              , (<> map (`PBftSigner` key) preW)  . dlPreW'
+              , (<> map (`PBftSigner` key) inW)   . dlInW'
+              ))
+          (id, id, id, id)
+          signerSlots
+
+    dlToSeq :: ([PBftSigner c] -> [PBftSigner c]) -> StrictSeq (PBftSigner c)
+    dlToSeq = Seq.fromList . sortOn pbftSignerSlotNo . ($ [])
+
+    preAnchor  = dlToSeq dlPreA
+    postAnchor = dlToSeq dlPostA
+    preWindow  = dlToSeq dlPreW
+    inWindow   = dlToSeq dlInW
+
+    counts = computeCounts inWindow
+
+encodePBftState :: (PBftCrypto c, Serialise (PBftVerKeyHash c))
                 => PBftState c -> Encoding
-encodePBftState st@PBftState{..} =
-    encodeVersion serializationFormatVersion0 $ mconcat [
```

Comment:
As mentioned in https://github.com/input-output-hk/ouroboros-network/issues/2383#issuecomment-672892991, I'm no longer bothering with `Ouroboros.Consensus.Util.Versioned`.